### PR TITLE
Set Pod phase to Failed if init container gets OOMKilled but exitCode…

### PR DESCRIPTION
**Pull request by ruiwen-zhao**
… is 0

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
When an init container gets OOM killed but the init process exits with 0, the state of the init container is correctly set as `Terminated`, but the pod status is incorrectly set as `Pending`. What should happen is that the pod status is set to `Failed`, so that pod GC can do its work to cleanup failed pods.

See also a previous bug fix (https://github.com/kubernetes/kubernetes/pull/104650) that checks OOMKilled status of init containers to decide whether the main containers should be started.

#### Which issue(s) this PR fixes:

Fixes #
https://github.com/kubernetes/kubernetes/issues/116676

Before this PR, deploying the image in https://github.com/kubernetes/kubernetes/issues/104454 results in pod stuck in Pending:

```
# k describe pod oomkill-init
Name:         oomkill-init
...
Status:       Pending
...
Init Containers:
  stress-init:
...
    State:          Terminated
      Reason:       OOMKilled
      Exit Code:    0
      Started:      Thu, 04 May 2023 20:32:30 +0000
      Finished:     Thu, 04 May 2023 20:32:31 +0000
```
after this PR, the pod is marked as Failed:
```
# k describe pod oomkill-init
Name:         oomkill-init
...
Status:       Failed
...
Init Containers:
  stress-init:
 ...
    State:          Terminated
      Reason:       OOMKilled
      Exit Code:    0
      Started:      Thu, 04 May 2023 20:35:01 +0000
      Finished:     Thu, 04 May 2023 20:35:02 +0000
```

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -1395,7 +1395,7 @@ func getPhase(pod *v1.Pod, info []v1.ContainerStatus, podIsTerminal bool) v1.Pod
 		case containerStatus.State.Running != nil:
 			pendingInitialization++
 		case containerStatus.State.Terminated != nil:
-			if containerStatus.State.Terminated.ExitCode != 0 {
+			if containerStatus.State.Terminated.ExitCode != 0 || containerStatus.State.Terminated.Reason == "OOMKilled" {
```
Comment (saschagrunert) [inline pkg/kubelet/kubelet_pods.go:1398]: I think we should define a const for `OOMKilled` somewhere.

**Review by saschagrunert — COMMENTED** [review]
Changes LGTM, just a nit about the const.

File: pkg/kubelet/kubelet_pods.go (hunk 1 above)
Comment (bobbypage) [reply pkg/kubelet/kubelet_pods.go:1398]: Can we re-use `isInitContainerFailed` here?

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/helpers.go#L149

**Comment by ruiwen-zhao** [comment]
Actually, the issue this PR tries to fix cannot be reproduced on 1.27, and can only be reproduced on 1.26. See https://github.com/kubernetes/kubernetes/issues/116676#issuecomment-1532198404. 

I am yet to find the change that fixed the exit code in container status between 1.26 and 1.27, but I think this PR is probably not needed anymore.

**Comment by bart0sh** [comment]
@ruiwen-zhao 
> I am yet to find the change that fixed the exit code in container status between 1.26 and 1.27, but I think this PR is probably not needed anymore.

Would it still make sense to fix this for 1.26 and 1.25?

**Comment by bart0sh** [comment]
If the issue is already fixed we shouldn't fix it again, I believe.
It would make sense to create PRs for 1.26 and 1.25 only.

**Comment by ruiwen-zhao** [comment]
> If the issue is already fixed we shouldn't fix it again, I believe. It would make sense to create PRs for 1.26 and 1.25 only.

Sorry do you mean backporting the big https://github.com/kubernetes/kubernetes/pull/115331 to 1.26 and 1.25, or making this PR a cherry-pick to 1.26 and 1.25? 

If the latter, I didn't know we were allowed to create cherry-picks without merging to main branch.

**Comment by ruiwen-zhao** [comment]
Since this issue is fixed in 1.27 by https://github.com/kubernetes/kubernetes/pull/115331. But https://github.com/kubernetes/kubernetes/pull/115331 is probably too big to be backported. I created https://github.com/kubernetes/kubernetes/pull/117788 to backport the fix to only 1.26.

**Note that the plan is to only backport this to 1.26 and 1.25, without merging to 1.27 and main branch.**

**Comment by bart0sh** [comment]
@ruiwen-zhao 
> Note that the plan is to only backport this to 1.26 and 1.25, without merging to 1.27 and main branch.

Would it make sense to close this PR then?

**Comment by ruiwen-zhao** [comment]
Yes lemme close this PR. If closing this PR somehow breaks the backport process, I will reopen it.

**Pull request closed without merging**